### PR TITLE
PSP: implement SDL_GetWindowSurface() API for direct VRAM access

### PR DESCRIPTION
**Pull request by rofl0r**
instead of dealing with renderers and textures, this allows you to create a window, set the pixel format using SDL_SetWindowDisplayMode() - preferably BGR565 for optimal speed - and then request SDL_GetWindowSurface() for each redraw, which provides you a surface with direct framebuffer access - that's why you need to call it on every repaint, rather than storing the surface variable.

note that you need to overwrite all pixels of the PSP visible area (480x272) to not encounter old data.

after you're done writing your pixels, call SDL_UpdateWindowSurface() and your changes will be sent to the graphics chip.

the result is a framerate of 60+ fps with BGR565 mode, unlike the other methods involving renderer and textures and a lot of copying.

# Existing Issue(s)
#9260

### CODE REVIEW

File: src/video/psp/SDL_pspvideo.c [hunk 1]
```diff
@@ -29,14 +29,119 @@
...
+    PSP_RenderData *rd;
+    SDL_WindowTextureData *data = SDL_GetWindowData(window, SDL_WINDOWTEXTUREDATA);
+    if(!data || !data->renderer || !window->surface) return -1;
+    rd = data->renderer->driverdata;
+    PSP_RenderPresent(data->renderer);
+    window->surface->pixels = rd->backbuffer;
+    return 0;
+}
+
+void PSP_DestroyWindowFramebuffer(_THIS, SDL_Window *window) {}
```
Comment (fjtrujy) [inline src/video/psp/SDL_pspvideo.c:144]: Shouldn't be this implemented?

File: src/video/psp/SDL_pspvideo.c [hunk 2]
```diff
@@ -29,14 +29,119 @@
...
+
+    return 0;
+}
+
+int PSP_UpdateWindowFramebuffer(_THIS, SDL_Window *window, const SDL_Rect *rects, int numrects) {
+    PSP_RenderData *rd;
+    SDL_WindowTextureData *data = SDL_GetWindowData(window, SDL_WINDOWTEXTUREDATA);
+    if(!data || !data->renderer || !window->surface) return -1;
+    rd = data->renderer->driverdata;
+    PSP_RenderPresent(data->renderer);
```
Comment (fjtrujy) [inline src/video/psp/SDL_pspvideo.c:139]: Can not be done in a different way? to expose `PSP_RenderPresent` looks a code smell, is not possible to do something as:
```c
data->renderer->RenderPresent(data->renderer)
```

**Comment by fjtrujy** [comment]
Good stuff by the way!

File: src/video/psp/SDL_pspvideo.c (hunk 1 above)
Comment (rofl0r) [reply src/video/psp/SDL_pspvideo.c:144]: i'd say that nobody on PSP creates more than one window - is it even possible ? - and the window lives as long as the program. so adding the destroy function would only help to have a 100% clean memory leak stats in valgrind on program termination. but i may be wrong.

File: src/video/psp/SDL_pspvideo.c (hunk 2 above)
Comment (rofl0r) [reply src/video/psp/SDL_pspvideo.c]: good catch!

**Review by fjtrujy — APPROVED** [review]
👏

**Comment by sharkwouter** [comment]
Hey @rofl0r thanks for implementing this, this is a really cool change. I'll do some testing with it tonight and merge it.

**Comment by slouken** [comment]
Rather than reaching into the PSP renderer internals, you should set properties on the renderer that will make it possible for this approach to be used by applications as well. The direct3d12 renderer uses this approach to expose the command queue in `SDL_PROP_RENDERER_D3D12_COMMAND_QUEUE_POINTER` for example.

Also, please remove the change using the texture access as a mask. If this is really necessary we should evaluate changing that field to a Uint32 mask for SDL3, and that can be done as a separate PR.

**Comment by rofl0r** [comment]
> Also, please remove the change using the texture access as a mask. If this is really necessary we should evaluate changing that field to a Uint32 mask for SDL3, and that can be done as a separate PR.

oops sorry, somehow that crept in. fixed now

> Rather than reaching into the PSP renderer internals, you should set properties on the renderer that will make it possible for this approach to be used by applications as well.

you mean getting the front/back pointers from the renderer ? not sure what you mean with applications. /me off reading the d3d12 code...

**Comment by slouken** [comment]
For SDL2 it's fine to leave as-is. Can you port this to SDL3 and use properties in that PR?

**Comment by slouken** [comment]
> hmm, it seems SDL_GetProperty is an SDL3-only feature. how do you want me to tackle this for the SDL2 branch ? is it OK if i create a property enum and a special function to query it inside PSP_render.c ?

Actually yeah, that seems like a better approach than splitting apart the PSP rendering internals to a separate header.

**Comment by sharkwouter** [comment]
Do you have some example code for testing this? I'm looking at this now, but I'm not sure how to implement it in my code.

**Comment by rofl0r** [comment]
> Can you port this to SDL3 and use properties in that PR?

Certainly, will start working on it as soon as this is merged

> Actually yeah, that seems like a better approach than splitting apart the PSP rendering internals to a separate header.

done

> Do you have some example code for testing this? I'm looking at this now, but I'm not sure how to implement it in my code.

i made sure to put all relevant info the commit message, but maybe this not-yet-cleaned-up diff for gnuboy using the new method helps more:

```diff
diff --git a/sys/sdl2/sdl2.c b/sys/sdl2/sdl2.c
index 9013348..340fff7 100644
--- a/sys/sdl2/sdl2.c
+++ b/sys/sdl2/sdl2.c
@@ -18,6 +18,17 @@
 #include "input.h"
 #include "rc.h"
 
+#if 0 && defined(__psp__)
+extern int sceMpegRingbufferConstruct(void *r, int i, void*p, int s, void*c, void* pa);
+/* shows n in the PPSSPP debug output */
+static void ppsspp_debug_int(int n) {
+        sceMpegRingbufferConstruct(0, n, 0, 0, 0, 0);
+}
+#else
+#include <stdio.h>
+static void ppsspp_debug_int(int n) { dprintf(2, "%d\n", n); }
+#endif
+
 extern void sdljoy_process_event(SDL_Event *event);
 
 struct fb fb;
@@ -25,10 +36,14 @@ struct fb fb;
 static int fullscreen = 0;
 static int use_altenter = 1;
 static int vsync;
+static int softfb;
 
 static SDL_Window *win;
 static SDL_Renderer *renderer;
 static SDL_Texture *texture;
+static SDL_Surface *fbs;
+
+static float stretchh, stretchv;
 
 static int vmode[3] = { 0, 0, 32 };
 
@@ -38,6 +53,8 @@ rcvar_t vid_exports[] =
 	RCV_VECTOR("vmode", &vmode, 3, "video mode: w h bpp"),
 	RCV_BOOL("fullscreen", &fullscreen, "start in fullscreen mode"),
 	RCV_BOOL("altenter", &use_altenter, "alt-enter can toggle fullscreen"),
+	RCV_FLOAT("stretchh", &stretchh, "horizontal stretch factor"),
+	RCV_FLOAT("stretchv", &stretchv, "vertical stretch factor"),
 	RCV_END
 };
 
@@ -72,12 +89,26 @@ void vid_init()
 	SDL_PixelFormat *format;
 	SDL_RendererInfo info;
 
+        softfb = 0;
+
+	stretchh = rc_getfloat("stretchh");
+	stretchv = rc_getfloat("stretchv");
+
 	if (!vmode[0] || !vmode[1])
 	{
 		if (scale < 1) scale = 1;
 		vmode[0] = 160 * scale;
 		vmode[1] = 144 * scale;
 	}
+#ifdef __psp__
+	vmode[0] = 480;
+	vmode[1] = 272;
+	vmode[2] = 16;
+	stretchh = 480 / 160;
+	stretchv = 272.0 / 144.0;
+	softfb = 1;
+	fullscreen = 1;
+#endif
 	if (vmode[2] == 32)
 		fmt = SDL_PIXELFORMAT_BGRA32;
 	else if(vmode[2] == 16)
@@ -99,6 +130,17 @@ void vid_init()
 		SDL_WINDOWPOS_UNDEFINED, vmode[0], vmode[1], flags)))
 		die("SDL: can't set video mode: %s\n", SDL_GetError());
 
+	if(softfb) {
+		SDL_DisplayMode dm = {.format = fmt, .w = vmode[0], .h = vmode[1] };
+		if(SDL_SetWindowDisplayMode(win, &dm) < 0)
+			die("SDL_SetWindowDisplayMode failed %s\n", SDL_GetError());
+		// SDL_BlitScaled
+		if(!(fbs = SDL_CreateRGBSurfaceWithFormat(0, dm.w, dm.h, vmode[2], fmt)))
+			die("SDL_CreateRGBSurfaceWithFormat failed %s\n", SDL_GetError());
+		pixels = fbs->pixels;
+		pitch = fbs->pitch;
+	} else {
+
 	/* for SDL2, which uses OpenGL, we internally use scale 1 and
 	   render everything into a 32bit high color buffer, and let the
 	   hardware do the scaling; thus "fb.delegate_scaling" */
@@ -122,10 +164,12 @@ void vid_init()
 	texture = SDL_CreateTexture(renderer, fmt,
 			SDL_TEXTUREACCESS_STREAMING, 160, 144);
 
-	SDL_ShowCursor(0);
+	SDL_LockTexture(texture, NULL, &pixels, &pitch);
+	}
+
+	//SDL_ShowCursor(0);
 
 	format = SDL_AllocFormat(fmt);
-	SDL_LockTexture(texture, NULL, &pixels, &pitch);
 
 	fb.delegate_scaling = 1;
 	fb.w = 160;
@@ -142,9 +186,12 @@ void vid_init()
 	fb.cc[2].r = format->Bloss;
 	fb.cc[2].l = format->Bshift;
 
-	SDL_UnlockTexture(texture);
+	if(!softfb)
+		SDL_UnlockTexture(texture);
+
 	SDL_FreeFormat(format);
 
+	ppsspp_debug_int(1);
 	fb.enabled = 1;
 	fb.dirty = 0;
 }
@@ -226,20 +273,41 @@ void vid_settitle(char *title)
 	SDL_SetWindowTitle(win, title);
 }
 
+static int drawtime;
+
 void vid_begin()
 {
 	void *p = 0;
 	int pitch;
-	SDL_LockTexture(texture, NULL, &p, &pitch);
+	if(!softfb)
+		SDL_LockTexture(texture, NULL, &p, &pitch);
+	else
+		p = fbs->pixels;
 	fb.ptr = p;
+	drawtime = SDL_GetTicks();
 }
 
 void vid_end()
 {
-	SDL_UnlockTexture(texture);
-	if (fb.enabled) {
-		SDL_RenderCopy(renderer, texture, NULL, NULL);
-		SDL_RenderPresent(renderer);
+	SDL_Rect destrect = {0, 0, 160*stretchh, 144*stretchv}, *d;
+	static const SDL_Rect srcrect = {0, 0, 160, 144};
+	if(stretchh > 0.0) d = &destrect;
+	else d = NULL;
+
+	if(!softfb) {
+		SDL_UnlockTexture(texture);
+		if (fb.enabled) {
+			SDL_RenderCopy(renderer, texture, d, d);
+			SDL_RenderPresent(renderer);
+		}
+	} else if(fb.enabled) {
+		SDL_Surface *screen;
+		if(!(screen = SDL_GetWindowSurface(win)))
+			die("SDL_GetWindowSurface failed %s\n", SDL_GetError());
+		SDL_BlitScaled(fbs, &srcrect, screen, d);
+		SDL_UpdateWindowSurface(win);
+		drawtime = SDL_GetTicks() - drawtime;
+		ppsspp_debug_int(drawtime);
 	}
 }
 
```
the old method is used if softfb == 0, the new one if softfb == 1

**Comment by rofl0r** [comment]
@slouken would it be acceptable to backport SDL properties to SDL2 ? if so i could use the same code for both branches, and introduce a property to tell the renderer that a texture is already swizzled (or do you see a better way to achieve this ?). when reading about the speed benefits of PSP swizzling, it seems there's a great speedup to be achieved if the SDL user already prepares the image data in a swizlled form.

**Comment by slouken** [comment]
No, the properties is a massive API change and can't be backported to SDL2. If you'd like to make this an SDL3 only feature, that's fine. We're wrapping up the API changes for SDL3 preview release soon, so the timing works out well.

**Comment by slouken** [comment]
It seems like PSP textures should always convert to swizzled form when being updated, for speed. Conceptually the application can't assume anything about the data backing the texture, it might be on the video card, or tiled or compressed for optimal usage in rendering.

**Comment by sharkwouter** [comment]
I found that this change does not break existing functionality, but would need an example of how to use the code to test more.

**Comment by rofl0r** [comment]
updated PR with a few SDL3-style fixes and rebased on current SDL2 HEAD.

**Review by slouken — APPROVED** [review]

**Comment by ccawley2011** [comment]
Would it be possible for applications to detect if the video driver is using double/triple buffering? This seems like a feature that would be useful on RISC OS as well.

Also, would it be necessary for the framebuffer to be lockable if the pointer to the pixel data can change, or would it be sufficient to explicitly document that the pointer may change after calling `SDL_UpdateWindowSurface()`?

**Comment by rofl0r** [comment]
> would it be necessary for the framebuffer to be lockable if the pointer to the pixel data can change

in the case of the PSP not, because its OS implements cooperative multitasking, no real hardware threads. this has caused me quite some headache already before i found that out, because i've been waiting for the sound thread to do its work, but nothing happened. the OS thread scheduler only kicks in on blocking syscalls. (for platforms like this, SDL should actually have a function like SDL_yield(), because abusing SDL_Delay(1) for this purpose already slims down the available time budget for 60 fps considerably).

> would it be sufficient to explicitly document that the pointer may change after calling SDL_UpdateWindowSurface()

that would certainly make sense, yes

**Review by sharkwouter — APPROVED** [review]
I just tested the example you sent and it works like expected. Existing code is also not impacted by this change. Thanks for this work, from my side this is approved.

**Comment by slouken** [comment]
@rofl0r, can you update your PR with any documentation that's needed? Once you've done that, I'll go ahead and merge this.

Also, I'm about to create an SDL 3.1.0 preview release. Do you want this in for that?

**Comment by rofl0r** [comment]
> can you update your PR with any documentation that's needed?

will do it in a few mins

> Also, I'm about to create an SDL 3.1.0 preview release. Do you want this in for that?

the getwindowsurface api already exists, so i guess it's nbd if the fast version for PSP is missing for the preview.

**Comment by rofl0r** [comment]
@slouken does this look alright ?

**Comment by ccawley2011** [comment]
> note that you need to overwrite all pixels of the PSP visible area (480x272) to not encounter old data.

Just to confirm, is it still possible to implement dirty rectangle based rendering this way by copying the modified areas for the last 2 frames instead of 1?

**Comment by rofl0r** [comment]
> Just to confirm, is it still possible to implement dirty rectangle based rendering this way by copying the modified areas for the last 2 frames instead of 1?

practically yes (i think), but that's highly specific to the system. though you could probably measure in your code after how many calls the pixels pointer is restored to the first value, so you know whether double or triple-buffering is at play (if at all).

**Comment by slouken** [comment]
> @slouken does this look alright ?

Looks great, thanks!

**Pull request merged**